### DISCUSSION TITLE
Workflows V1

### DISCUSSION THREAD

**Discussion by elviskahoro** [warpdotdev/Warp#625] [Product roadmap]
# [0] Overview and Mockups

Entering commands in the terminal can be tedious and hard–developers often need to construct commands one argument and flag at a time, often using StackOverflow.

Workflows (formerly called Tasks) make it easy to browse, search, execute and share commands (or a series of commands)--without needing to leave your terminal.

Warp wants to make it easier to enter commands by launching with 100+ common workflows and an open-source file format--so devs can create private workflows for themselves or their team.

Please comment with feedback, we’ll create an individual sub-thread for each mockup and memo section!

|[A] The Workflow’s menu.|
|:--:|
| ![A](https://user-images.githubusercontent.com/29553206/150206227-e5ae4eec-b322-4509-aba8-f67ee23eab39.png) |

|[B] Filling a Workflow in the Input Editor with the menu expanded.|
|:--:|
| ![B](https://user-images.githubusercontent.com/29553206/150206255-2288326d-4c5c-459a-b1fc-77c046c13bc7.png) |

|[C] Filling a Workflow in the Input Editor with the menu collapsed and tab completion.|
|:--:|
| ![C](https://user-images.githubusercontent.com/29553206/150207821-d0d02b9f-686e-475a-b1fe-a3741fb3f707.png) |

# [1] Pain Points and Goals

Our team, tends to run into a few pain points when entering commands:

1. We need to find a command but
	1. the arguments and flags do not represent the task we are trying to perform
		1. To “undo last git commit” -> `git reset --hard SOFT`
	2. our team's playbook (guide) lives outside of the terminal 
2. We can’t find aliases that other team members have made
	1. No store to browse aliases
	2. Sharing, revoking access, or updating aliases through version control is not straightforward, especially
	   since aliases are global
3. The shell does not provide enough documentation or UI to help enter parameters to commands

# [2] How is this Different from Aliases?

Power users tend to save aliases, create shell functions, and leverage CLI tools that streamline this.
Aliases, however, have major pain points:

1. need to context switch
	1. leave vim, source dotfiles, or reset shell
2. it’s difficult to attach documentation
3. are not easily searchable or sharable
4. are not easily parameterized

Getting aliases and functions to a productive state requires an upfront investment that’s justifiable for devs who
spend most of their workday in the terminal but less so for beginners and casual users.

# [3] Workflows V1

We’re planning on the first launch of Workflows to include:

1. The ability to search and execute a workflow directly from the input editor
2. The top 100 workflows across various commands already bundled
   1. `git`, `sed`, and `grep`
3. The ability for devs to create private workflows in an open-source file-format, including repo-specific workflows that live in a `.warp` directory
4. The ability to create a workflow from a Block by right-clicking on it

# [4] S/O Open Source Projects

These are some of the open source projects that we’ve gotten inspiration from: [tldr;](https://github.com/tldr-pages/tldr) [cheat.sh](https://github.com/chubin/cheat.sh), [cheatsheets](https://github.com/rstacruz/cheatsheets), [navi](https://github.com/denisidoro/navi), [howdoi](https://github.com/gleitz/howdoi), [cheat](https://github.com/cheat/cheat), [how2](https://github.com/santinic/how2), [kb](https://github.com/gnebbia/kb), [eg](https://github.com/srsudar/eg).

Navi [in particular!](https://github.com/denisidoro/navi/blob/master/docs/cheatsheet_syntax.md)

# [5] Question for the Community

We’ve made a thread for each section; please reply in the appropriate threads to help organize discussion!

### We'll ship with some prebuilt workflows, what are some workflows to include? 
If you can, try to use this format:

#### Title of workflow (what does it do)

Extract a field from a json

#### Command

`cat channel_versions.json | jq '.stable.versions'`

#### Parameters

`channel_versions.json` -> a json file

#### Articulation or links to StackOverflow (or other sites)

https://stackoverflow.com/questions/39228500/extract-a-specific-field-from-json-output-using-jq

**Comment by elviskahoro** [comment]
# [5] We'll ship with some prebuilt workflows, what are some workflows to include?

If you can, try to use this format:

## Title of workflow (what does it do)
Extract a field from a json

## Command
`cat channel_versions.json | jq '.stable.versions'`

## Parameters
`channel_versions.json` -> a json file

## Articulation or links to StackOverflow (or other sites)
https://stackoverflow.com/questions/70777851/need-feedback-on-warps-workflows-v1

**Reply by cliffordfajardo** [reply]
## 1.
Title of workflow: 
- restart mac audio service
Command:
- `sudo pkill -9 coreaudiod`
https://apple.stackexchange.com/a/366841


## 2
Title of workflow: 
- kill a port 
Command:
```
sudo lsof -i:8080 // this outputs the PID
kill -9 <PID Number>
```

**Reply by lavabyrd** [reply]
Title of workflow:

* Project start - open a particular directory, open 4 tiled windows (2x2 square) and run particular services in each
Command
* Dependance but similar to `cd ~/{project}`, split the terminal and run `./app/run1` in the first terminal, `cat app/file.log` in the second, etc.

**Comment by elviskahoro** [comment]
# [0] Overview and Mockups

Share your feedback on the mockups here! When referring to a specific mockup, try to include the mock's id / label: A, B, or C

**Reply by cliffordfajardo** [reply]
I like the command pallete for searching for tasks.

1. (Mock A)
Once a command is executed, it would be nice if there was a successful [toast banner ](https://web.dev/building-a-toast-component/)that informs the user that the command successfully executed

**Comment by elviskahoro** [comment]
# [1] Pain Points and Goals

Share your feedback and experiences on the pain points and our goals here!

**Reply by cliffordfajardo** [reply]
I could imagine Warb workflows being shared with others.
It would be nice to have a user profile with Warp (connect with Github?)
You can follow others.

Example
- Imagine my coworkers and I use Warp>
- Having the ability to follow someone (coworker etc) & be able to see their public scripts would be nice.
   - I work with Backend engineers & SRE's and they have all the awesome cool commands for SSH & grepping machines & it would be nice to be able to see those  (if they shared them publicly in Warp user intefacs)

**Reply by elviskahoro** [reply]
From brendan via Discord: https://discord.com/channels/851854972600451112/907407371997970492/907407376569737277

> There are many times I find myself running commands in an exploratory manner and then rerunning them in order later once I have it figured out. For e.g. building then deploying to preprod then running integ tests against preprod. The first time it might take a handful of times to get the syntax right, but after that I just repeat them in order. It can then be scripted but sometimes I can't be bothered and just keep repeating them.

> It would be really cool if warp could detect this and create a script/command out of it automatically. Not sure what the ideal UX would be, but one possibility would be a toast in the bottom right like how VS code does it. Hit accept and it will ask you to name it and confirm the steps involved

**Reply by elviskahoro** [reply]
From jfmomnty2 via Discord: https://discord.com/channels/851854972600451112/907407371997970492/907692365395935264
> This sounds really cool, especially if you could have an option to do it in an "ephemeral" way. My thinking is, often I find myself running the same 3-6 commands many times one day, but then never again b/c it was just some specific setup logic for a particular thing I was working on. 

> Would be really cool if Warp could detect the same sequence of commands and say "you've run these x commands n times in the last 20 minutes. Would you like to save them? And if so, save them forever or just for now?

**Reply by lavabyrd** [reply]
I'd love to have something similar to the project files in VSCode so you don't have to modify the path for each user, you just use a .warp file or similar that can be placed in the directory. This would work quite similarly to launch.json files for vscode

**Reply by elviskahoro** [reply]
Discord Comment Relating (feature-requests - 07/14/2021)
> "Something cool I think would be valuable to me is ad-hoc variable or alias creation and a UI to see what you have access to. For instance, I have several IPs that I regularly SSH into. I'd love to either save those IPs as variable, or even just set the entire command as an alias, without having to jump in and configure a file."

**Comment by elviskahoro** [comment]
# [3] Workflows V1

### What functionality should Workflows support?

**Reply by elviskahoro** [reply]
From mantis via Discord

>1. "Interesting. Such could be project aware. Such that if I build some rust or go project, Warp could then index such and I can build something without even having to cd to that folder. Such could be automatic. Like build this or that."
> 2. "And as you likely can guess, Homebrew support."
> 3. "Maybe also native platform commands like working with file systems apfs or zfs."

**Reply by elviskahoro** [reply]
From @ccpost via Discord:
https://discord.com/channels/851854972600451112/915391043812352021/916640242457542697

> Bit of a cart-before-the-horse thing since this is still in the initial phase, but I was thinking of 2 main ways you'd probably want to save / share these tasks:
> 1. I can see quite a lot of more-general things being useful in many contexts; globally, inside any git repository, inside any Go code tree. Those kind of things probably fit pretty neatly into being managed directly by Warp (saved into the cloud / shared with your team when Warp implements paid team collab / etc).
> 2. But I'd also get a lot of mileage out of tasks that are project-specific as a replacement for how projects usually end up with things in a Makefile or dev-scripts directory or similar. For those kind of things, I'd really want the definition of the task to be versioned inside the repo along with the code and not synced through an external Warp mechanism. Those kind of things tend to be real tied to other things happening in the repo, and you'd want to keep them in sync with the rest of the project for them to work properly. Ideally Warp would pick these up if the current directory or any parent contains like a .warptasks directory or something so that it'd still work if you're down inside the project tree. 
> In a way: like the GitHub Actions workflows in .github/workflows, but local and more interactive.

**Reply by elviskahoro** [reply]
> After sharing is implemented and Workflows can be loaded from the web (commands.dev) it'd be nice to have trust (this Workflow) functionality. ie some sub-commands in the workflow would need explicit approval (press Y) each time their ran until the users explicitly "trusts" the script (adds it to their inventory)

**Reply by elviskahoro** [reply]
From ｍｅｍｅ　ｍａｎ　維ぞ易 via Discord: https://discord.com/channels/851854972600451112/906269689116770304/935538790653628426
> When you implement workflows, it would be extremely useful to have a GUI for setting environment variables locally for the workflow so it makes it easier to configure actual environments faster.

**Reply by elviskahoro** [reply]
From CA_Lobo via Discord - https://discord.com/channels/851854972600451112/906350632263376916/906393591188963368
> I'd expand ephemer's idea to a library of preconfigured snippets that can be selected individuallly or chained/piped together... a command palette selection + pipe cmd + a second command palette selection or an exisiting snippet... add a textual UI user query to this mix cuz I'm always forgetting the right options for a script that I use maybe twice a month....  and perhaps a oneline description for each snippet as a reminder... user case is that most the stuff I do in the terminal window is repetive and needs to be automated...  and avoid typing rm -rf *  when I meant rm -rf *.build ...

**Reply by elviskahoro** [reply]
From CA_Lobo via Discord
> Excellent start! I like it! Glad to see others have the same ideas! Do we have a UI to go along with it?  Can the UI chain existing tasks together in the same principles of a Unix tools- eg multiple reusable individual tasks when chained together can automate a multi step process? something along the lines of "do a clean development checkout, do a debug build, package it, smoke test it and notify me and the QA team if passes, and any failures notify me" I can't wait to play with this as I have a dozen "scraps of paper" receipes on my computer to deal with repetitive tasks... humm, in a previous life, one thing I'd love to be able to do would be to share these tasks/receipes with others... so exporting/importing of packaged tasks would be something to think about...

**Reply by elviskahoro** [reply]
From Discord (feature-requests - 08/08/2021)
> "That's perfect! I'm imagining being able to build up a selection of these blocks, and add them into a playlist - to automate console tasks a little like how ansible does"

**Reply by elviskahoro** [reply]
Sourced from Discord: (General Channel - 09/03/2021)
> "It'd also be super cool to be able to reference executed blocks (either the command to be run live again or using existing output) in other blocks/commands and have visual cues while doing so."

**Reply by denisidoro** [reply]
It would be nice if there was a list of possible values for a given variable. For example, for a variable such as `docker_container_id`, a list with `2fa1b` and `c4d2a` could be shown. 

This list could have multiple columns showing extra data.

**Reply by denisidoro** [reply]
Syntax highlighting

**Comment by elviskahoro** [comment]
# [4] Open Source Projects

### Are there other open source projects related to workflows that we should be aware of?
The one's we already know of: [tldr;](https://github.com/tldr-pages/tldr) [cheat.sh](https://github.com/chubin/cheat.sh), [cheatsheets](https://github.com/rstacruz/cheatsheets), [navi](https://github.com/denisidoro/navi), [howdoi](https://github.com/gleitz/howdoi), [cheat](https://github.com/cheat/cheat), [how2](https://github.com/santinic/how2), [kb](https://github.com/gnebbia/kb), [eg](https://github.com/srsudar/eg).

### What opportunities are there for collaboration?

**Reply by cliffordfajardo** [reply]
Other workflow launchers
- https://www.scriptkit.com/. The [creator's  twitter ](https://twitter.com/johnlindquist?ref_src=twsrc%5Egoogle%7Ctwcamp%5Eserp%7Ctwgr%5Eauthor)has lots of great demos

Raycast
- I like their API for defining workflows & UIs
- they're app is a Swift app, but their extension ecosystem uses React. [Docs](https://developers.raycast.com/)
   - Pros of react+typescript API: lots of engineers know javascript & react. React is declarative
   - React has an API called  `react-reconciler` which allows you to create your own declarative abstraction over imperative code.
      - For example: with reconciler you can have users write React/Typescript code & you could output to whatever platform you want. This is how react-native, react-three-fiber & [react-blessed](https://github.com/Yomguithereal/react-blessed) (terminal renderer) and many other libraries use react, the difference is they choose a different rendering targer

**Reply by lavabyrd** [reply]
The ability to trigger a workflow from something like https://www.raycast.com/ so you could utilize both Raycast for spotlight style workflow running but also run the workflows directly if needs be

**Reply by denisidoro** [reply]
Author of navi here. I'd be more than glad to integrate navi with Warp if possible :)

For example, navi could support the open-source file-format you're proposing:
```
The ability for devs to create private workflows in an open-source file-format, including repo-specific workflows that live in a .warp directory
```

**Reply by elviskahoro** [reply]
From vinaygattu via Discord:
https://discord.com/channels/851854972600451112/906269689116770304/949611067363983411
> Any integration with Alfred workflows and warp

**Reply by pomdtr** [reply]
I quite like raycast script commands syntax: https://github.com/raycast/script-commands
It is simple, allows to pass arguments, and works with any language.
An important feature for me is that raycast script are still executable as regular scripts.
